src/mame/konami/hornet.cpp

```diff
@@ -1260,7 +1333,7 @@ void hornet_state::sscope(machine_config &config)
 	m_voodoo[1]->set_status_cycles(1000); // optimization to consume extra cycles when polling status
 	m_voodoo[1]->set_screen("rscreen");
 	m_voodoo[1]->set_cpu(m_dsp[1]);
-	m_voodoo[1]->vblank_callback().set_inputline(m_maincpu, INPUT_LINE_IRQ1);
+	m_voodoo[1]->vblank_callback().set([this] (bool c) { m_sscope_irq1_triggered = true; });
```

Comment:
This setup cannot possibly work: you are forcing an irq1 on both states of the voodoo vblank callback. 
I smell that Voodoo is somehow eating the callback instead, cfr. https://github.com/mamedev/mame/blob/3b9fa5f3c69ea498be27bc74d3c9fc25f9b6bde6/src/devices/video/voodoo.cpp#L2748 , where in particular m_vsyncstop is always zero and rscreen definitely isn't set as 0 for minimum vpos here. Also, I see a CG Control Register that ties irq1 ack to bit 7, is it cleared inside a single frame if at all?

If you are unsure you may as well post the relevant error.log here, in particular the voodoo rscreen setup.

Reply:
Checked since I was bored: everytime it executes an irq (on gameplay at least) it acks BOTH lines with 0xc0 / 0xd0 writes. I suspect that whatever is EXRES0 / EXRES1 is NOT irq ack, and this has been cargo culted to nwk-tr.cpp and zr107.cpp drivers.

Reply:
This doesn't trigger IRQ1 by itself. The original assumption in the driver was that the scope screen was tied to the vblank but I do not see how that is possible considering it takes 57 calls to IRQ1 for a single screen update to occur. The callback here is only used as a known "good" timing (as in, by the time the vblank is called, calling IRQ1 won't cause things to crash). Once this gets called at least once then the timer function will start asserting IRQ1.

This is what's logged in the console, but I am not sure if this is what you're after.
```
[:voodoo1] hSync=74-941, bp=106, vis=768  vSync=3-259, bp=17, vis=236
[:voodoo1] Horiz: 106-873 (1017 total)  Vert: 17-252 (262 total) -- 
[:voodoo1] Standard resolution, 60.114504 Hz
[:voodoo1] yoffs: 17 vsyncstart: 259 vsyncstop: 0

[:voodoo0] hSync=39-603, bp=42, vis=512  vSync=6-422, bp=27, vis=384
[:voodoo0] Horiz: 42-553 (644 total)  Vert: 27-410 (428 total) -- 
[:voodoo0] Medium resolution, 58.411215 Hz
[:voodoo0] yoffs: 27 vsyncstart: 422 vsyncstop: 0
```

For what it's worth, I did a bit of testing to make sure that the vblank callback was being called as I expected and I didn't see any issues with the vblank callback. If you think you see the issue with the logic here then please do look into the code more because I couldn't find anything obviously wrong.

I just threw together this small func to print out the timings:
```
m_voodoo[1]->vblank_callback().set([this] (bool c) {
	if (c) {
		m_sscope_irq1_triggered = true;
		printf("voodoo[1] vblank callback called %lf (diff: %lf)\n", machine().time().as_double(), (machine().time() - m_last_time).as_double());
		m_last_time = machine().time();

		m_maincpu->set_input_line(INPUT_LINE_IRQ1, ASSERT_LINE);
	}
});
```

Output:
```
voodoo[1] vblank callback called 46.587570 (diff: 0.016635)
voodoo[1] vblank callback called 46.604205 (diff: 0.016635)
voodoo[1] vblank callback called 46.620840 (diff: 0.016635)
voodoo[1] vblank callback called 46.637475 (diff: 0.016635)
voodoo[1] vblank callback called 46.654110 (diff: 0.016635)
voodoo[1] vblank callback called 46.670745 (diff: 0.016635)
voodoo[1] vblank callback called 46.687380 (diff: 0.016635)
voodoo[1] vblank callback called 46.704015 (diff: 0.016635)
voodoo[1] vblank callback called 46.720650 (diff: 0.016635)
voodoo[1] vblank callback called 46.737285 (diff: 0.016635)
voodoo[1] vblank callback called 46.753920 (diff: 0.016635)
voodoo[1] vblank callback called 46.770555 (diff: 0.016635)
voodoo[1] vblank callback called 46.787190 (diff: 0.016635)
voodoo[1] vblank callback called 46.803824 (diff: 0.016635)
voodoo[1] vblank callback called 46.820459 (diff: 0.016635)
voodoo[1] vblank callback called 46.837094 (diff: 0.016635)
voodoo[1] vblank callback called 46.853729 (diff: 0.016635)
voodoo[1] vblank callback called 46.870364 (diff: 0.016635)
voodoo[1] vblank callback called 46.886999 (diff: 0.016635)
voodoo[1] vblank callback called 46.903634 (diff: 0.016635)
```

>Also, I see a CG Control Register that ties irq1 ack to bit 7, is it cleared inside a single frame if at all?

I am skeptical if that information is even correct to be honest. Neither IRQ0 or IRQ1's function seem to do something that looks like clearing the bits based on the IRQ called. Bits 6 *and* 7 always seem to be set (from boot), regardless of whether the IRQ has been called or not.

Reply:
> I am skeptical if that information is even correct to be honest. Neither IRQ0 or IRQ1's function seem to do something that looks like clearing the bits based on the IRQ called. Bits 6 and 7 always seem to be set (from boot), regardless of whether the IRQ has been called or not.

That's what I've said in my last comment. Voodoo has it's own assert / clear logic, so by the time irq0 asserts (which is **faster** than irq1) then irq1 clears and never gets executed.
I've also compiled by just removing the custom acks, and yes the scope refreshes fine with that only.

Reply:
Well I'll be damned. That actually does seem to fix it. I assumed the IRQ clear lines were correct until you asked and I looked into it while writing my last comment. 

I'm going to give the rest of the Hornet library a run through to see if it fixes anything else or at the last doesn't break anything before I commit the changes. Thanks a bunch! Never would've suspected that to have been the issue.

Reply:
I'm seeing some regressions with Teraburst and NBA Play By Play. I'll remove the scope screen hack from the PR since I have at least a better direction to go with it now, and send another PR for the scope screen after I try to figure out why Teraburst and NBA Play By Play are acting weird now. Teraburst's insert coin text is blinking a million times a second, and NBA Play By Play freezes when inserting a coin.

Reply:
Both games have no IRQ1 callback from what I can tell, they all runs on IRQ0 only. The alternative is that irq pending/clear mechanism should be smarter than just chaining vblank_callback() to a `set_inputline`, that is "effectively ack the line if the irq was taken" to sysregs_w $7, and assert on `state = true`. I don't have enough PowerPC proficiency to tell however, and I guess there's no vector here?

Reply:
Teraburst handles IRQ 0 and 2 (IRQ handler func is at 0x80000508). NBA Play By Play handles IRQ 0 and 1 (handler func at 0x80140508) but IRQ 1 looks like it should never be called or else it'll throw you into an infinite loop while writing to the 7seg on the PCB(?).

The only consistent thing I found for clearing the IRQ is that IRQ0 and IRQ1 both write 0 to 0x74000012 and 0x74000014 respectively, which are registers on the k037122 chip (the chip for 2D graphics that's on the same board as the Voodoo).

Reply:
Yeah, because `k037122_device` also handles CRTC, so it may drive the irq ack as well (cfr. K053252 CCU). To be sure I'd monitor how Voodoo is setup, particularly the fbiInit1 and fbiInit5 registers since those also controls who drives VCLK (if internal or external).

Reply:
Here's all of the writes on boot for NBA Play By Play.
```
[:voodoo0] fbiinit3: 0010c001
[:voodoo0] fbiinit1: 00201102
[:voodoo0] fbiinit0: 00000416
[:voodoo0] fbiinit0: 00000412
[:voodoo0] fbiinit0: 00000410
[:voodoo0] fbiinit0: 00000411
[:voodoo0] fbiinit1: 10032102
[:voodoo0] fbiinit2: 80000040
[:voodoo0] fbiinit3: 0010c001
[:voodoo0] fbiinit4: 00000001
[:voodoo0] fbiinit1: 10032100
[:voodoo0] fbiinit4: 00000001
[:voodoo0] fbiinit1: 10032108
[:voodoo0] fbiinit2: 80000040
[:voodoo0] fbiinit4: 00000003
[:voodoo0] fbiinit2: 18000040
[:voodoo0] fbiinit2: 18400040
[:voodoo0] fbiinit2: 186000e0
[:voodoo0] fbiinit2: 186300e0
[:voodoo0] fbiinit1: 10033108
[:voodoo0] fbiinit1: 10033108
[:voodoo0] fbiinit2: 186300e0
[:voodoo0] fbiinit1: 10033188
[:voodoo0] fbiinit3: 0010c033
[:voodoo0] fbiinit3: 5fd0c033
[:voodoo0] fbiinit4: 00000003
[:voodoo0] fbiinit4: 07fc0003
[:voodoo0] fbiinit4: 07fd2003
[:voodoo0] fbiinit4: 07fd2063
[:voodoo0] fbiinit0: 00000751
[:voodoo0] fbiinit0: 6c400351
[:voodoo0] fbiinit1: 10032188
[:voodoo0] fbiinit3: 5fd0a033
[:voodoo0] fbiinit1: 10032088
[:voodoo0] fbiinit2: 186300e0
[:voodoo0] fbiinit0: 6c400350
[:voodoo0] fbiinit1: 10032088
[:voodoo0] fbiinit0: 6c402350
```

And here's one for Silent Scope
```
[:voodoo1] fbiinit3: 0010c001
[:voodoo1] fbiinit1: 00201102
[:voodoo1] fbiinit0: 00000416
[:voodoo1] fbiinit0: 00000412
[:voodoo1] fbiinit0: 00000410
[:voodoo1] fbiinit0: 00000411
[:voodoo1] fbiinit1: 10032102
[:voodoo1] fbiinit2: 80000040
[:voodoo1] fbiinit3: 0010c001
[:voodoo1] fbiinit4: 00000001
[:voodoo1] fbiinit1: 10032100
[:voodoo1] fbiinit4: 00000001
[:voodoo1] fbiinit1: 10032108
[:voodoo1] fbiinit2: 80000040
[:voodoo1] fbiinit4: 00000003
[:voodoo1] fbiinit2: 18000040
[:voodoo1] fbiinit2: 18400040
[:voodoo1] fbiinit2: 186000e0
[:voodoo1] fbiinit2: 186300e0
[:voodoo1] fbiinit1: 10032188
[:voodoo1] fbiinit1: 10032188
[:voodoo1] fbiinit2: 1862d0e0
[:voodoo1] fbiinit1: 100321c8
[:voodoo1] fbiinit3: 0010c033
[:voodoo1] fbiinit3: 3ad0c033
[:voodoo1] fbiinit4: 00000003
[:voodoo1] fbiinit4: 07fc0003
[:voodoo1] fbiinit4: 07fd0e03
[:voodoo1] fbiinit4: 07fd0e63
[:voodoo1] fbiinit0: 00000751
[:voodoo1] fbiinit0: 6c400351
[:voodoo1] fbiinit1: 100321c8
[:voodoo1] fbiinit3: 3ad0a033
[:voodoo1] fbiinit1: 100320c8
[:voodoo1] fbiinit2: 1862d0e0
[:voodoo1] fbiinit0: 6c400350
[:voodoo1] fbiinit1: 100320c8
[:voodoo1] fbiinit0: 6c402350
[:voodoo0] fbiinit3: 0010c001
[:voodoo0] fbiinit1: 00201102
[:voodoo0] fbiinit0: 00000416
[:voodoo0] fbiinit0: 00000412
[:voodoo0] fbiinit0: 00000410
[:voodoo0] fbiinit0: 00000411
[:voodoo0] fbiinit1: 10032102
[:voodoo0] fbiinit2: 80000040
[:voodoo0] fbiinit3: 0010c001
[:voodoo0] fbiinit4: 00000001
[:voodoo0] fbiinit1: 10032100
[:voodoo0] fbiinit4: 00000001
[:voodoo0] fbiinit1: 10032108
[:voodoo0] fbiinit2: 80000040
[:voodoo0] fbiinit4: 00000003
[:voodoo0] fbiinit2: 18000040
[:voodoo0] fbiinit2: 18400040
[:voodoo0] fbiinit2: 186000e0
[:voodoo0] fbiinit2: 186300e0
[:voodoo0] fbiinit1: 10032188
[:voodoo0] fbiinit1: 10032188
[:voodoo0] fbiinit2: 186300e0
[:voodoo0] fbiinit1: 10032188
[:voodoo0] fbiinit3: 0010c033
[:voodoo0] fbiinit3: 5fd0c033
[:voodoo0] fbiinit4: 00000003
[:voodoo0] fbiinit4: 07fc0003
[:voodoo0] fbiinit4: 07fd2003
[:voodoo0] fbiinit4: 07fd2063
[:voodoo0] fbiinit0: 00000751
[:voodoo0] fbiinit0: 6c400351
[:voodoo0] fbiinit1: 10032188
[:voodoo0] fbiinit3: 5fd0a033
[:voodoo0] fbiinit1: 10032088
[:voodoo0] fbiinit2: 186300e0
[:voodoo0] fbiinit0: 6c400350
[:voodoo0] fbiinit1: 10032088
[:voodoo0] fbiinit0: 6c402350
```

It does seem to have drive_video_timing, drive_video_dclk, and video_timing_vclk bits set if I'm reading this correctly.

Reply:
fbiinit1 bit 17 means vclk slave input select. That's the most likely setting for an external source driving everything else, including irqs.

Reply:
Ok. I guess this all ties into what I was researching about this code the other day about involving changing the clock settings via sysreg_w 7 (the 0x0c comment I added). 

The Voodoo code wants to hijack the screen parameters no matter what when it's in use and tbh I do not feel comfortable making significant changes to the Voodoo code to stop that, so I will implement everything up but not including actually configuring screen parameters via the k037122. That should make it easy to calculate the frame period and trigger a callback for asserting and clearing the IRQ based on the k037122 instead of the Voodoo. I already did quite a bit of work on figuring out how games that can switch clocks do so, and how they set the CRTC parameters (Gradius 4 is the only Hornet game that had CRTC parameters for a 640x480 31kHz mode, but it breaks video display when you actually use it).

Reply:
I attempted to rewrite the code to use the k037122 as the main source and but there were still regressions with games running too fast (Silent Scope's in-game timers are ticking too fast for example). I've opted to remove the code that clears IRQ1 in sysreg_w and added a note about it because removing the IRQ1 clear doesn't seem to have any noticeable regressions in any of the games I tested but does improve the Silent Scope scope screen to be mostly playable without hacks.